core/src/test/java/org/apache/iceberg/rest/TestHTTPClient.java

```diff
@@ -0,0 +1,250 @@
+import static org.mockserver.model.HttpRequest.request;
+import static org.mockserver.model.HttpResponse.response;
+
+/***
+ * Exercises the RESTClient interface, specifically over a mocked-server using the actual HttpRESTClient code.
+ */
+public class TestHTTPClient {
+
+  private static final int port = 1080;
+  private static final String uri = String.format("http://127.0.0.1:%d", port);
+  private static final JsonFactory FACTORY = new JsonFactory();
+  private static final ObjectMapper MAPPER = new ObjectMapper(FACTORY);
+
+  private static ClientAndServer mockServer;
+  private static RESTClient restClient;
+
+  @BeforeClass
+  public static void beforeClass() {
+    MAPPER.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+    mockServer = startClientAndServer(port);
+    restClient = HTTPClient
+        .builder()
+        .uri(uri)
+        .mapper(MAPPER)
+        .withBearerAuth("auth_token")
+        .build();
+  }
+
+  @AfterClass
+  public static void stopServer() throws IOException {
+    mockServer.stop();
+    restClient.close();
+  }
+
+  @Test
+  public void testPostSuccess() throws Exception {
+    testHttpMethodOnSuccess(HttpMethod.POST);
+  }
+
+  @Test
+  public void testPostFailure() throws Exception {
+    testHttpMethodOnFailure(HttpMethod.POST);
+  }
+
+  @Test
+  public void testGetSuccess() throws Exception {
+    testHttpMethodOnSuccess(HttpMethod.GET);
+  }
+
+  @Test
+  public void testGetFailure() throws Exception {
+    testHttpMethodOnFailure(HttpMethod.GET);
+  }
+
+  @Test
+  public void testDeleteSuccess() throws Exception {
+    testHttpMethodOnSuccess(HttpMethod.DELETE);
+  }
+
+  @Test
+  public void testDeleteFailure() throws Exception {
+    testHttpMethodOnFailure(HttpMethod.DELETE);
+  }
+
+  @Test
+  public void testHeadSuccess() throws JsonProcessingException {
+    testHttpMethodOnSuccess(HttpMethod.HEAD);
+  }
+
+  @Test
+  public void testHeadFailure() throws JsonProcessingException {
+    testHttpMethodOnFailure(HttpMethod.HEAD);
+  }
+
+  public static void testHttpMethodOnSuccess(HttpMethod method) throws JsonProcessingException {
+    Item body = new Item(0L, "hank");
+    String asJson = MAPPER.writeValueAsString(body);
+    AtomicInteger errorCounter = new AtomicInteger(0);
+    Consumer<ErrorResponse> onError = (error) -> {
+      errorCounter.incrementAndGet();
+      throw new RuntimeException("Failure response");
+    };
+
+    // Using different paths Keeps the expectations unique for the test's mock server
+    String path = String.format("%s_success", method);
+
+    HttpRequest mockRequest = request("/" + path).withMethod(method.name().toUpperCase(Locale.ROOT));
+    if (method.usesRequestBody()) {
+      mockRequest = mockRequest.withBody(asJson);
+    }
+
+    HttpResponse mockResponse = response().withStatusCode(200);
+    if (method.usesResponseBody()) {
+      mockResponse = mockResponse.withBody(asJson);
+    }
+
+    mockServer
+        .when(mockRequest)
```

Comment:
It looks like this test validates the request payload because if it is not present then this `when` does not match. Is that correct?

Reply:
That's correct. The mockserver is essentially storing all of these as assertion cases. Then, when running the tests, it will attempt to match in 3 steps (with all needing to pass):

1) Check that the method matches, e.g .`GET`, `HEAD`, etc.
2) Check that the path matches that of the request.
3) Check the request payload matches against what was sent.

So it is correct to say that this validates the request payload.